### PR TITLE
Add Test.Data.Conversions.Integral.testTree

### PR DESCRIPTION
**Pull request by epicallan**
* [x] Depends on #246

Followup: 

* Add `Test.Data.Conversions.Textual.testTree` with text, bytestring & string conversion tests

### CODE REVIEW

File: conversions/src/Data/Conversions.hs [hunk 1]
```diff
@@ -61,8 +61,12 @@ instance (MonadError (UserBoundError Int64 Natural) m) => Conversion (m Natural)
   convert = convertErrorFromIntegral
 
 instance (MonadError (UserBoundError Integer Text) m) => Conversion (m Natural) Integer where
-  convert value = maybe (throwError $ UserBoundError value "0" "Natural") pure
-    $ checkedFromIntegral value
+  convert value = do
+    when (value < 0) $ throwError userBoundError
```
Comment (epicallan) [inline conversions/src/Data/Conversions.hs:65]: Comment: This is a fix, previously we would get an impure error.

File: conversions/src/Data/Conversions.hs [hunk 2]
```diff
@@ -136,6 +140,21 @@ instance Conversion Int Word8 where
 instance Conversion Int Word16 where
   convert = fromIntegral
 
+instance Conversion Int Word32 where
+  convert = fromIntegral
+
+-- do something better
+instance (MonadError (BoundError Word64 Int) m) => Conversion (m Int) Word64 where
+  convert value = do
+    when (value > convertImpure (maxBound @Int)) -- TODO: standardize into helper and re-use
```
Comment (epicallan) [inline conversions/src/Data/Conversions.hs:149]: instance of https://github.com/mbj/mhs/pull/243/commits/900e4ebf304c13d2143ce6721f8b2ab787002618#r1231396417

File: conversions/src/Data/Conversions.hs [hunk 3]
```diff
@@ -136,6 +140,21 @@ instance Conversion Int Word8 where
...
+      $ throwError boundError'
+
+    maybe (throwError boundError') pure $ checkedFromIntegral value
+    where
+      boundError' :: BoundError Word64 Int
+      boundError' = BoundError value
+
+
```
Comment (epicallan) [inline conversions/src/Data/Conversions.hs:157]: TODO: remove extra line

File: conversions/src/Data/Conversions.hs [hunk 4]
```diff
@@ -136,6 +140,20 @@ instance Conversion Int Word8 where
 instance Conversion Int Word16 where
   convert = fromIntegral
 
+instance Conversion Int Word32 where
+  convert = fromIntegral
+
+-- do something better
+instance (MonadError (BoundError Word64 Int) m) => Conversion (m Int) Word64 where
+  convert value = do
+    when (value > convertImpure (maxBound @Int))
```
Comment (epicallan) [inline conversions/src/Data/Conversions.hs:149]: instance of https://github.com/mbj/mhs/pull/243/files#r1231396417

File: conversions/src/Data/Conversions.hs [hunk 5]
```diff
@@ -169,11 +199,19 @@ instance (MonadError (BoundError Integer Word16) m) => Conversion (m Word16) Int
 instance (MonadError (BoundError Integer Word8) m) => Conversion (m Word8) Integer where
   convert = convertBoundedFromIntegral
 
+instance (MonadError (BoundError Integer Word) m) => Conversion (m Word) Integer where
+  convert = convertBoundedFromIntegral
+
 instance (MonadError (BoundError Int Word64) m) => Conversion (m Word64) Int where
-  convert = checkedFromIntegralToBounded
+  convert value = do
+    when (value < 0) $ throwError boundError'
```
Comment (epicallan) [inline conversions/src/Data/Conversions.hs:207]: instance of https://github.com/mbj/mhs/pull/243/files#r1231396417

File: conversions/src/Data/Conversions/Hedgehog.hs [hunk 6]
```diff
@@ -0,0 +1,110 @@
+module Data.Conversions.Hedgehog
+  ( BiConvert(..)
+  , Generators(..)
+  , mkConvertEither
+  , run
```
Comment (epicallan) [inline conversions/src/Data/Conversions/Hedgehog.hs:5]: Comment: I plan to use `Data.Conversions.Hedgehog` outside current `mhs/conversions` tests

File: conversions/test/Test/Data/Conversions/Integral.hs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+         )
+      => (Hedgehog.Range integer -> Hedgehog.Gen integer)
+      -> Tasty.TestTree
+    toNatural genInteger
+      = Test.run biConvert generators
+      where
+        generators :: Test.Generators integer
+        generators = Test.Generators
+          { success = [genInteger range]
+          , failure = [pure $ minBound @integer]
```
Comment (dkubb) [inline conversions/test/Test/Data/Conversions/Integral.hs:58]: __suggestion:__ It looks like this is missing the exact boundary check for integer, which would be `-1`.

File: conversions/package.yaml [hunk 8]
```diff
@@ -3,16 +3,20 @@ _common/package: !include "../common/package.yaml"
...
 dependencies:
 - base
 - bytestring
 - exceptions
+- hedgehog
+- mprelude
 - mtl
 - scientific
+- tasty
+- tasty-hedgehog
```
Comment (mbj) [inline conversions/package.yaml:19]: Also likely test dependencies?

File: conversions/test/Test/Data/Conversions/Integral.hs [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+         )
+      => (Hedgehog.Range word -> Hedgehog.Gen word)
+      -> Tasty.TestTree
+    toWord genSource
+      = run @word generators
+      where
+        generators :: Test.Generators Natural
+        generators = Test.Generators
+          { success = [convert <$> genSource range]
+          , failure = [pure (convert $ maxBound @Word64)]
```
Comment (dkubb) [inline conversions/test/Test/Data/Conversions/Integral.hs:117]: __suggestion:__ Sort of along the same lines as my [previous statement](https://github.com/mbj/mhs/pull/243/files#r1231447485) could we specify a range that is *just* above the boundary up to `maxBound @Word64`?

File: conversions/src/Data/Conversions/Hedgehog.hs (hunk 6 above)
Comment (mbj) [reply conversions/src/Data/Conversions/Hedgehog.hs:5]: gotcha if we are re-exporting disregard my comment about reducing library dependencies.

File: conversions/test/Test/Data/Conversions/Integral.hs (hunk 9 above)
Comment (epicallan) [reply conversions/test/Test/Data/Conversions/Integral.hs]: I made your suggested change which allowed me to eliminate a standalone test for `Word64`

File: conversions/test/Test/Data/Conversions/Integral.hs (hunk 7 above)
Comment (epicallan) [reply conversions/test/Test/Data/Conversions/Integral.hs]: > Just demonstrating that -1 is invalid is probably fine. wdyt?


I agree and using `-1` resonates well with our principle of using the simplest primitive. `-1` is the simplest in this case.

File: conversions/src/Data/Conversions/Hedgehog.hs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+          runGeneratorList
+            :: Foldable f
+            => (source -> PropertyIO)
+            -> f (Hedgehog.Gen source)
+            -> Hedgehog.Property
+          runGeneratorList assertion = Hedgehog.property . traverse_ runGenerator
+            where
+              runGenerator :: Hedgehog.Gen source -> PropertyIO
+              runGenerator generator = assertion =<< Hedgehog.forAll generator
```
Comment (dkubb) [inline conversions/src/Data/Conversions/Hedgehog.hs:75]: __suggestion: (nitpick)__ Could this be written as `rrunGenerator = (assertion =<<) . Hedgehog.forAll`?

Comment (epicallan) [reply conversions/src/Data/Conversions/Hedgehog.hs:75]: Agreed. I will make this change

File: conversions/src/Data/Conversions/Hedgehog.hs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+                <> show expectation
+                <> " but received "
+                <> show value
+                <> " for conversion from "
+                <> show source
+
+mkConvertEither :: Show error => (a -> Either error b) -> (a -> Either ConversionError b)
+mkConvertEither f = left (ConversionError . show) . f
+
+typeName :: forall (a :: Type). Typeable a => String
```
Comment (dkubb) [inline conversions/src/Data/Conversions/Hedgehog.hs:101]: __suggestion: (nitpick)__ There's a space missing before the `.`.

File: conversions/src/Data/Conversions/Hedgehog.hs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            -> PropertyIO
+          roundTrip expectation source = case expectation of
+            Success -> onSuccess $ toTarget source
+            Failure -> onFailure $ toTarget source
+            where
+              onFailure :: Either ConversionError target -> PropertyIO
+              onFailure = either (const (pure ())) mkError
+
+              onSuccess :: Either ConversionError target -> PropertyIO
+              onSuccess = either mkError (either mkError (const (pure ())) . fromTarget)
```
Comment (dkubb) [inline conversions/src/Data/Conversions/Hedgehog.hs:86]: __suggestion: (nitpick)__ Should the function definitions be flipped to match the order in the case statement?

**Review by dkubb — APPROVED** [review]

File: conversions/src/Data/Conversions/Hedgehog.hs (hunk 12 above)
Comment (epicallan) [reply conversions/src/Data/Conversions/Hedgehog.hs:86]: Yes agreed

**Pull request merged**